---
title: "Linear_Lasso_WGCNA_SoftThreshold_16"
author: 
- name: "Computational Immunogenomics Core (CIG Core)"
  affiliation: "Center for Systems Immunology, University of Pittsburgh"
  email: "sis186@pitt.edu"
date: "`r Sys.Date()`"
output: 
  html_document:
    toc: true # table of content true
    toc_depth: 3  # upto three depths of headings (specified by #, ## and ###)
    toc_float:
      collapsed: true
      smooth_scroll: true
    number_sections: true  ## if you want number sections at each table header
    theme: "default" # many options for theme, this one is my favorite.
    highlight: tango  # specifies the syntax highlighting style
---

```{r setup, include=FALSE, echo=FALSE}
knitr::opts_chunk$set(echo = FALSE, warning = FALSE, message = FALSE)


## Libraries
suppressMessages({
  library(openxlsx)
  library(biomaRt)
  library(stringr)
  library(matrixStats)
  library(pheatmap)
  library(dplyr)
  library(ggfortify)
  library(org.Hs.eg.db)
  library(WebGestaltR)
  library(DESeq2)
  library(apeglm)
  library(EnhancedVolcano)
  library('janitor') #use the function: row_to_names
})
library(ROCR) #AUC
library(cvTools) #AUC
library("RNAseqQC")
library("ensembldb")
library("dplyr")
library("ggplot2")
library("purrr")
library("tidyr")
library("tibble")
library("magrittr")
library(DEGreport)
library(writexl)
library("glmnet")
library("caret")  # For cv.glmnet function
## Sourcing
source("/ix/cigcore/utils/code/de.heatmap.R")
library(brglm)
library(tximport)
library(sleuth)
library(dplyr)
library(VennDiagram)
library(AnnotationDbi)
library(enrichplot)
library(tidyverse)
library(cowplot) #plot_grid
library(glmnet)
library(doParallel)
registerDoParallel(cores = 4) 
library(nestedcv)
## Set paths
proj_path <- "/ix/cigcore/proj/biswas"
data_path <- file.path(proj_path, "data", "Ascherman_TotalRNA_07_20_2023")
resPath <- gsub(pattern="data", replacement="result", data_path)
experiment <- "4-11-Lasso_WGCNA_16_linear"
experimentPath <- file.path(resPath, experiment)
dir.create(experimentPath, showWarnings = TRUE, recursive = TRUE)
#metadata_path <- file.path(data_path, "metadata_remove_unpaired_low_align.csv")
#dir.create(webgestalt_path, recursive=TRUE)
metadata_path <- file.path(data_path, "all_metadata_remove_unpaired_low_align.csv") 
file_path <- file.path(data_path, "TPM_ordered_variance.csv")
```


```{r count, include=FALSE, echo=FALSE}
coldata <- read.csv(file.path(metadata_path), row.names = 1) #Metadata
coldata$sample <- rownames(coldata)  # new variable:sample = rownames in 'coldata'
#coldata <- coldata[ , c("sample", "treatment")]
coldata$treatment <- ifelse(coldata$treatment == "Pre", 0, 1)
coldata$treatment_numeric <- as.numeric(coldata$treatment)
# coldata$treatment_probit <- qnorm(coldata$treatment_numeric)  --> the probabilities of the binary outcome onto the standard normal distribution --> give me inf/-inf 
```

```{r, echo=FALSE}
TPM_ordered_variance <- read.csv(file.path(file_path), row.names = 1)
```



# WGCNA result

> Soft threshold = 16

soft threshold = **16**
	
![](/ix/cigcore/proj/biswas/result/Ascherman_TotalRNA_07_20_2023/4-1-2023-WGCNA_SoftThreshold_16/thumbnail_image.png){width=60%}.

> Modules = 29

**29** modules in total

![](/ix/cigcore/proj/biswas/result/Ascherman_TotalRNA_07_20_2023/3-25-2023-WGCNA/module_16.png){width=70%}.

|  Positive modules | Spearman correlation (p-value) | 
| ------------- |:-------------:|
|  **lightgreen (152 genes)** | 0.14 (0.1)  | 


|  Negative modules | Spearman correlation (p-value) | 
| ------------- |:-------------:|
|  **darkred (63 genes)** | -0.12 (0.2)  | 
|  **midnightblue (303 genes)** | -0.1 (0.2) | 
 



```{r, out.width = '70%'}
knitr::include_graphics("/ix/cigcore/proj/biswas/result/Ascherman_TotalRNA_07_20_2023/4-1-2023-WGCNA_SoftThreshold_16/heatmap_SoftThreshold_16.png")
```

```{r, echo=FALSE}
#X: expression matrix
expression_matrix <- as.matrix(TPM_ordered_variance)
#Y: response variable 
response_variable <- coldata$treatment_numeric
#response_variable <- as.factor(response_variable)
```

```{r, echo=FALSE}
file_path_lightgreen <- "/ix/cigcore/proj/biswas/result/Ascherman_TotalRNA_07_20_2023/3-25-2023-WGCNA/module_csv_SoftThreshold_16/lightgreen.csv"
file_path_darkred <- "/ix/cigcore/proj/biswas/result/Ascherman_TotalRNA_07_20_2023/3-25-2023-WGCNA/module_csv_SoftThreshold_16/darkred.csv"
file_path_midnightblue <- "/ix/cigcore/proj/biswas/result/Ascherman_TotalRNA_07_20_2023/3-25-2023-WGCNA/module_csv_SoftThreshold_16/midnightblue.csv"

positive <- read.csv(file_path_lightgreen) #152
negative_1 <- read.csv(file_path_darkred) #63
negative_2 <- read.csv(file_path_midnightblue) #303

positive_expression_matrix <- expression_matrix[,positive$gene] #lightgreen: 152
negative_expression_matrix_1 <- expression_matrix[,negative_1$gene] #darkred: 63
negative_expression_matrix_2 <- expression_matrix[,negative_2$gene] #midnightblue: 303
```

# Run LASSO on treatment-positive modules (Module lightgreen: 152 genes)

* Alpha = 1

* Nested cross validation

     + outer loop method: leave-one-out
     + inner loop method: leave-one-out
 
```{r}
  alpha <- 1
  
  set.seed(123)
  
   n_outer_folds <- length(response_variable)  # LOOCV
    n_inner_folds <- 23 # LOOCV
    outer_folds <- lapply(1:length(response_variable), function(i) i)
    
    set.seed(123)
    nested_cv_result <- nestcv.glmnet(
      y = response_variable, 
      x = positive_expression_matrix, 
      family = "gaussian",
      outer_method = "LOOCV", 
      n_inner_folds = n_inner_folds,
      outer_folds = outer_folds, 
      alphaSet = 1,
      finalCV = TRUE
    )
    #cv.glmnet: pass "standardize TRUE" 
    
    
    # Get the lambda value, model and non-zero coefficients selected during cross-validation
    best_lambda <- nested_cv_result$final_param["lambda"]
    best_lambda_value <- as.numeric(best_lambda)
    current_model <- nested_cv_result$final_fit
    non_zero_coefficients <- rownames(nested_cv_result[["final_coef"]])[-1]
  
  cat("Tuned lambda value:\n", best_lambda_value, "\n")
  current_model
  cat("Non-zero Coefficients:\n", non_zero_coefficients, "\n")

  gene_positive <- non_zero_coefficients
  

```

## RSEM and R-squared
```{r}
nested_cv_result[["summary"]]
```

## List of genes with non-zero coefficients (11 Genes)

```{r}

ensemblID <- gene_positive
attributes <- c("ensembl_gene_id", "external_gene_name")
ensemblMart <- useMart("ensembl", dataset="hsapiens_gene_ensembl")
external_gene_names_positive <- getBM(attributes = attributes, filters = "ensembl_gene_id", values = gene_positive, mart = ensemblMart)

knitr::kable(external_gene_names_positive, "html") %>%
  kableExtra::kable_styling(full_width = FALSE)

```




# Run LASSO on treatment-negative modules (Module darkred: 63 genes)

* Alpha = 1

* Nested cross validation

     + outer loop method: leave-one-out
     + inner loop method: k-fold; k = **4**
     + Didn't find any non-zero coeffients when using leave-one-out for inner loop
 
```{r}
  alpha <- 1
  
  set.seed(123)
  
   n_outer_folds <- length(response_variable)  # LOOCV
    n_inner_folds <- 4
    outer_folds <- lapply(1:length(response_variable), function(i) i)
    
    set.seed(123)
    nested_cv_result <- nestcv.glmnet(
      y = response_variable, 
      x = negative_expression_matrix_1, 
      family = "gaussian",
      outer_method = "LOOCV", 
      n_inner_folds = n_inner_folds,
      outer_folds = outer_folds, 
      alphaSet = 1,
      finalCV = TRUE
    )
    
    # Get the lambda value, model and non-zero coefficients selected during cross-validation
    best_lambda <- nested_cv_result$final_param["lambda"]
    best_lambda_value <- as.numeric(best_lambda)
    current_model <- nested_cv_result$final_fit
    non_zero_coefficients <- rownames(nested_cv_result[["final_coef"]])[-1]
  
  cat("Tuned lambda value:\n", best_lambda_value, "\n")
  current_model
  cat("Non-zero Coefficients:\n", non_zero_coefficients, "\n")

  gene_negative_1 <- non_zero_coefficients
  
```



## RSEM and R-squared
```{r}
nested_cv_result[["summary"]]
```

## List of genes with non-zero coefficients  (4 genes) 

```{r}
ensemblID <- gene_negative_1
attributes <- c("ensembl_gene_id", "external_gene_name")
#ensemblMart <- useMart("ensembl", dataset="hsapiens_gene_ensembl")
external_gene_names_negative_1 <- getBM(attributes = attributes, filters = "ensembl_gene_id", values = gene_negative_1, mart = ensemblMart)


external_gene_names_negative_1$external_gene_name <- ifelse(external_gene_names_negative_1$external_gene_name == "", 
                                                     external_gene_names_negative_1$ensembl_gene_id, 
                                                   external_gene_names_negative_1$external_gene_name)


knitr::kable(external_gene_names_negative_1, "html") %>%
  kableExtra::kable_styling(full_width = FALSE)
```



# Run LASSO on treatment-negative module (Module midnightblue: 303 genes)

* Alpha = 1

* Nested cross validation


     + outer loop method: leave-one-out
     + inner loop method: k-fold; k = **4**
     + Didn't find any non-zero coeffients when using leave-one-out for inner loop
 
```{r}
  alpha <- 1
  
  set.seed(123)

  
   n_outer_folds <- length(response_variable)  # LOOCV
    n_inner_folds <- 4
    outer_folds <- lapply(1:length(response_variable), function(i) i)
    
    set.seed(123)
    nested_cv_result <- nestcv.glmnet(
      y = response_variable, 
      x = negative_expression_matrix_2, 
      family = "gaussian",
      outer_method = "LOOCV", 
      n_inner_folds = n_inner_folds,
      outer_folds = outer_folds, 
      alphaSet = 1,
      finalCV = TRUE
    )
    
    # Get the lambda value, model and non-zero coefficients selected during cross-validation
    best_lambda <- nested_cv_result$final_param["lambda"]
    best_lambda_value <- as.numeric(best_lambda)
    current_model <- nested_cv_result$final_fit
    non_zero_coefficients <- rownames(nested_cv_result[["final_coef"]])[-1]
  
  cat("Tuned lambda value:\n", best_lambda_value, "\n")
  current_model
  cat("Non-zero Coefficients:\n", non_zero_coefficients, "\n")

  gene_negative_2 <- non_zero_coefficients
  
```

## RSEM and R-squared
```{r}
nested_cv_result[["summary"]]
```


## List of genes with non-zero coefficients  (1 gene)

```{r}
ensemblID <- gene_negative_2
attributes <- c("ensembl_gene_id", "external_gene_name")
#ensemblMart <- useMart("ensembl", dataset="hsapiens_gene_ensembl")
external_gene_names_negative_2 <- getBM(attributes = attributes, filters = "ensembl_gene_id", values = gene_negative_2, mart = ensemblMart)


external_gene_names_negative_2$external_gene_name <- ifelse(external_gene_names_negative_2$external_gene_name == "", 
                                                     external_gene_names_negative_2$ensembl_gene_id, 
                                                   external_gene_names_negative_2$external_gene_name)


knitr::kable(external_gene_names_negative_2, "html") %>%
  kableExtra::kable_styling(full_width = FALSE)
```



# Final Lasso nested cv (Total 16 genes)

```{r}
total_gene <- rbind(external_gene_names_positive, external_gene_names_negative_1, external_gene_names_negative_2)
```

```{r}
alpha <- 1

final_model_matrix <- as.data.frame(expression_matrix[, total_gene$ensembl_gene_id])
final_model_matrix <- merge(as.data.frame(final_model_matrix), as.data.frame(coldata[1]), by = "row.names", all.x = TRUE)
cv_matrix <- expression_matrix[, total_gene$ensembl_gene_id]

genes_final_model <- paste(total_gene$ensembl_gene_id, collapse = " + ")

formula_str <- paste("treatment ~", genes_final_model)

   n_outer_folds <- length(final_model_matrix$treatment)  # LOOCV
    n_inner_folds <- 23
    outer_folds <- lapply(1:length(response_variable), function(i) i)
    
    set.seed(123)
    final_lasso_nested_cv <- nestcv.glmnet(
      y = final_model_matrix$treatment, 
      x = cv_matrix, #'x' must be numeric
      family = "gaussian",
      outer_method = "LOOCV", 
      n_inner_folds = n_inner_folds,
      outer_folds = outer_folds, 
      alphaSet = alpha,
      finalCV = TRUE
    )
    

    best_lambda <- final_lasso_nested_cv$final_param["lambda"]
    best_lambda_value <- as.numeric(best_lambda)
    current_model <- final_lasso_nested_cv$final_fit
    non_zero_coefficients <- rownames(final_lasso_nested_cv[["final_coef"]])[-1]
  
  cat("Tuned lambda value:\n", best_lambda_value, "\n")
  current_model
  cat("Non-zero Coefficients:\n", non_zero_coefficients, "\n")
  gene_final <- non_zero_coefficients
```

```{r}
final_lasso_nested_cv[["final_coef"]]
```


## RSEM and R-squared in CV

```{r}
final_lasso_nested_cv[["summary"]]
```


## Final linear model (13 genes)

```{r}
final_model_matrix <- as.data.frame(expression_matrix[, gene_final])
final_model_matrix <- merge(as.data.frame(final_model_matrix), as.data.frame(coldata[5]), by = "row.names", all.x = TRUE)

genes_final_model <- paste(gene_final, collapse = " + ")

formula_str <- paste("treatment_numeric ~", genes_final_model)


finalModel_lm <- lm(formula_str,
                  data = data.frame(final_model_matrix))
                
#try linear link
summary(finalModel_lm)
```

## R-squared in final linear model
```{r}
summary(finalModel_lm)$adj.r.squared
```

# Pearson correlation

```{r}
library(Hmisc)
library(corrplot)
x <- lapply(final_model_matrix[,2:14], as.numeric)
y <- final_model_matrix$treatment_numeric

rcor.df <- rcorr(as.matrix(as.data.frame(lapply(final_model_matrix[,2:14], as.numeric))), y, type ="pearson")

par(mar=c(1,1,1,1))  
corrplot(rcor.df$r, method = 'number', number.cex = 0.7, tl.col = "black", col = COL2('RdBu', 6), width = 50, height = 50)

```

## Genes with significant correlation (p < 0.05)

Specialized the insignificant value according to the significant level, and hide the correlation with itself

```{r}
par(mar=c(1,1,1,1)) 
cor.df_R <- corrplot(rcor.df$r, method = "number", number.cex = 0.7, type = "upper", tl.col = "black", 
                    p.mat = rcor.df$P, sig.level = 0.05, diag = FALSE, width = 50, height = 50)


```

Genes with p value < 0.05

```{r}
p_values <- rcor.df[["P"]]["y", -length(rcor.df[["P"]])]
p_values <- p_values[-length(p_values)]

significant_genes <- p_values[p_values < 0.05]

print(significant_genes)
```


# Final genes list (Total 13 genes)


| Modules (size)       | Module correlation to treatment     | Genes selected by lasso |
| ------------- |:-------------:| -----:|
|  **lightgreen (152 genes)**      | Positive | 11 |
|  **darkred (63 genes)**      | Negative      |   1 |
| **midnightblue (303 genes)**      | Negative      |    1 |

  * Compared with logistic regression, gene ENSG00000240671 (IGKV1-8) is the additional gene 

  
```{r}
total_gene <- rbind(external_gene_names_positive, external_gene_names_negative_1, external_gene_names_negative_2)
final_gene_matrix <- total_gene[total_gene$ensembl_gene_id %in% gene_final, ]
rownames(final_gene_matrix) <- seq_len(nrow(final_gene_matrix))
knitr::kable(final_gene_matrix, "html",row.names = TRUE) %>%
  kableExtra::kable_styling(full_width = FALSE)
```



# Heatmap (Pre-treatment vs Post-treatment)


```{r}
library(RColorBrewer)
selected_genes <- final_gene_matrix$ensembl_gene_id
zMatrix <- t(TPM_ordered_variance)
gene_zMatrix <- zMatrix[rownames(zMatrix) %in% selected_genes, , drop = FALSE]
gene_zMatrix <- as.data.frame(gene_zMatrix)

breaksList = seq(-3, 3, by = 0.06)

sorted <- coldata[order(coldata$treatment == '1', coldata$treatment), ]
cluster_treatment <- as.data.frame(t(sorted[, c(1:4)]))

gene_zMatrix_reordered <- gene_zMatrix[, match(names(cluster_treatment), names(gene_zMatrix))]


condition_reordered <- data.frame(sorted[, 1])
colnames(condition_reordered)[1] <- "treatment"
rownames(condition_reordered) <- colnames(gene_zMatrix_reordered)

gene_zMatrix_reordered <- gene_zMatrix_reordered[match(final_gene_matrix$ensembl_gene_id, rownames(gene_zMatrix_reordered)), ]
rownames(gene_zMatrix_reordered) <- final_gene_matrix$external_gene_name

p1 <- pheatmap(gene_zMatrix_reordered,
               color = colorRampPalette(rev(brewer.pal(n = 7, name =
  "RdYlBu")))(100),
               show_rownames = TRUE, 
               cluster_cols = FALSE, 
               cluster_rows = FALSE, 
               annotation_col = condition_reordered, 
               annotation_colors = list(treatment = c("0" = "lightpink", "1" = "lightgreen")),
               cutree_rows = 2, 
               cutree_cols = 2, 
               fontsize_row = 7,
               breaks = breaksList,
               legend = TRUE
)

p1

```

## Heatmap for Log-FoldChange for each patient

```{r}
selected_genes <- final_gene_matrix$ensembl_gene_id
zMatrix <- t(TPM_ordered_variance)
gene_zMatrix <- zMatrix[rownames(zMatrix) %in% selected_genes, , drop = FALSE]
gene_zMatrix <- as.data.frame(t(gene_zMatrix))
df <- merge(as.data.frame(gene_zMatrix), as.data.frame(coldata), by = "row.names", all.x = TRUE)
df <- df[, c(2,3,4,5,6,7,8,9,10,11,12,13,14,15,17)]

df2 <- df %>%
  pivot_longer(cols = -c(patient_id, treatment), names_to = "Gene", values_to = "logTPM") %>%
  group_by(patient_id, Gene) 

treatment_1 <- df2 %>% dplyr::filter(treatment == 1)
treatment_0 <- df2 %>% dplyr::filter(treatment == 0)
merged_data <- merge(treatment_1, treatment_0, by = c("patient_id", "Gene"), suffixes = c("_treatment_1", "_treatment_0"))
merged_data <- merged_data %>%
  mutate(diff_logTPM = logTPM_treatment_1 - logTPM_treatment_0) %>%
  dplyr::select(Gene, patient_id, diff_logTPM) %>% 
  pivot_wider(names_from = Gene, values_from = diff_logTPM) 

rownames(merged_data) <- merged_data$patient_id
patient <- as.data.frame(unique(coldata$patient_id))
rownames(patient) <- patient$`unique(coldata$patient_id)`
merged_data_patinet_order <- merged_data[rownames(patient), ]

rownames(merged_data_patinet_order) <- merged_data_patinet_order$patient_id
gene_LFC <- as.data.frame(t(merged_data_patinet_order))
gene_LFC <- gene_LFC[-1, ]

gene_LFC_reordered <- gene_LFC[match(final_gene_matrix$ensembl_gene_id, rownames(gene_LFC)), ]
rownames(gene_LFC_reordered) <- final_gene_matrix$external_gene_name
#keep patient_id the same as the previous heatmap
breaksList = seq(-3, 3, by = 0.06)
p4 <- pheatmap(gene_LFC_reordered, 
         show_rownames = TRUE, 
         cluster_cols = FALSE, 
         cluster_rows = FALSE, 
         cutree_rows = 2, 
         cutree_cols = 2, 
         breaks = breaksList,
         fontsize_row = 7)
```

# Pathway Analysis ORA

```{r ora1, echo=FALSE, warning=FALSE, message=FALSE, include=FALSE}
library(clusterProfiler) # perform ORA
library(org.Hs.eg.db) # Homo sapiens annotation --> gene identifier conversion
library(ReactomePA) # Reactome

#Each gene is represented by a human Entrez gene ID, which is the default keytype used by the clusterProfiler functions (and the only keytype compatible with ReactomePA::enrichPathway).
cp_path <- file.path(experimentPath, "pathway_analysis_ORA")
dir.create(cp_path, recursive=TRUE)

genes <- final_gene_matrix$ensembl_gene_id
entrez_genes <- bitr(genes, fromType = "ENSEMBL", toType = "ENTREZID", OrgDb = "org.Hs.eg.db")

# Reactome pathway over-representation analysis
    reactome_1 <- enrichPathway(gene = entrez_genes$ENTREZID, organism = "human", pvalueCutoff=0.2, pAdjustMethod = "BH", qvalueCutoff = 0.2, readable=TRUE)
    ORA <- list()
    ORA[[paste0("Reactome", "_WGCNA_16_linear")]] <- reactome_1
    tab.reactome_1 <- as.data.frame(reactome_1)
    
# KEGG pathway over-representation analysis
    kegg_1 <- enrichKEGG(gene = entrez_genes$ENTREZID, organism = "hsa",  pvalueCutoff=0.2, pAdjustMethod = "BH", qvalueCutoff = 0.2, use_internal_data = FALSE)
    # mapping geneID to gene Symbol
    kegg_gene_1 <- setReadable(kegg_1, OrgDb = "org.Hs.eg.db", keyType="ENTREZID")
    kegg_gene_1@result$Description <- gsub(pattern = " - Homo sapiens (human)", replacement = "", kegg_gene_1@result$Description, fixed = T)    
    ORA[[paste0("KEGG", "_WGCNA_16_linear")]] <- kegg_gene_1
    tab.kegg_1 <- as.data.frame(kegg_gene_1)

    plots_list <- list()
    
for(a1 in names(ORA)){
  reacORkegg <- ORA[[a1]]
  if(nrow(reacORkegg) !=0){
    # Create directories for each result
    ORA_path <- file.path(cp_path, a1)
    dir.create(ORA_path, recursive=TRUE)
    
    ORA_df <- as.data.frame(reacORkegg)
    write.csv(ORA_df, file.path(ORA_path, paste0(a1, ".csv")))
    
    # Plot enrichment result - dotplot
    fit <- plot(dotplot(reacORkegg, showCategory = 40))
    png(filename = file.path(ORA_path, paste0(a1, "_dotplot.png")), res = 300, width = 17, height = 10, units = "in")
    print(fit)
    dev.off()
    
    plots_list[[a1]] <- fit
  }
  
  if (nrow(reacORkegg) > 20){ #change number
    # Weighted Set Cover of geneSets
    weightedPath <- file.path(cp_path, "weightedSetCover", a1)
    dir.create(weightedPath, recursive=TRUE)
    
    setCoverNum = abs(0.40*(nrow(ORA_df)))
    nThreads = 4
    idsInSet <- sapply(ORA_df$geneID, strsplit, split="/")
    names(idsInSet) <- ORA_df$ID    
    minusLogP <- -log(ORA_df$pvalue)
    minusLogP[minusLogP == Inf] <- -log(.Machine$double.eps)
    wscRes <- weightedSetCover(idsInSet=idsInSet, costs=(1 / minusLogP), topN=setCoverNum, nThreads=nThreads)

    wscRes_full <- ORA_df[c(match(wscRes$topSets, ORA_df$ID)),]
    wscRes_full <- wscRes_full[order(wscRes_full$p.adjust), ]
    write.csv(wscRes_full, file.path(weightedPath, paste0(a1, ".csv")))
    
    # Plot enrichment result of weighted set cover - dotplot
    reacORkegg2 <- reacORkegg
    reacORkegg2@result <- wscRes_full
      
    # Set larger bottom margin to make space for y-axis labels
    fit <- plot(dotplot(reacORkegg2, showCategory = 40), main = paste0(a1))
    png(filename = file.path(weightedPath, paste0(a1, "_dotplot.png")), res = 300, width = 17, height = 28, units = "in")
    print(fit)
    dev.off()
    plots_list[[paste0(a1, "weightSet")]] <- fit
  }
}    
```


> Reactome

![](/ix/cigcore/proj/biswas/result/Ascherman_TotalRNA_07_20_2023/4-11-Lasso_WGCNA_16_linear/pathway_analysis_ORA/weightedSetCover/Reactome_WGCNA_16_linear/Reactome_WGCNA_16_linear_dotplot.png).

> KEGG

![](/ix/cigcore/proj/biswas/result/Ascherman_TotalRNA_07_20_2023/4-11-Lasso_WGCNA_16_linear/pathway_analysis_ORA/KEGG_WGCNA_16_linear/KEGG_WGCNA_16_linear_dotplot.png).

