---
title: "Logistic_WGCNA_16"
author: 
- name: "Computational Immunogenomics Core (CIG Core)"
  affiliation: "Center for Systems Immunology, University of Pittsburgh"
  email: "sis186@pitt.edu"
date: "`r Sys.Date()`"
output: 
  html_document:
    toc: true # table of content true
    toc_depth: 3  # upto three depths of headings (specified by #, ## and ###)
    toc_float:
      collapsed: true
      smooth_scroll: true
    number_sections: true  ## if you want number sections at each table header
    theme: "default" # many options for theme, this one is my favorite.
    highlight: tango  # specifies the syntax highlighting style
---

```{r setup, include=FALSE, echo=FALSE}
knitr::opts_chunk$set(echo = FALSE, warning = TRUE, message = FALSE)

## Libraries
suppressMessages({
  library(openxlsx)
  library(biomaRt)
  library(stringr)
  library(matrixStats)
  library(pheatmap)
  library(dplyr)
  library(ggfortify)
  library(org.Hs.eg.db)
  library(WebGestaltR)
  library(DESeq2)
  library(apeglm)
  library(EnhancedVolcano)
  library('janitor') #use the function: row_to_names
})
library(ROCR) #AUC
library(cvTools) #AUC
library("RNAseqQC")
library("ensembldb")
library("dplyr")
library("ggplot2")
library("purrr")
library("tidyr")
library("tibble")
library("magrittr")
library(DEGreport)
library(writexl)
library("glmnet")
library("caret")  # For cv.glmnet function
## Sourcing
source("/ix/cigcore/utils/code/de.heatmap.R")
library(brglm)
library(tximport)
library(sleuth)
library(dplyr)
library(VennDiagram)
library(AnnotationDbi)
library(enrichplot)
library(tidyverse)
library(cowplot) #plot_grid
library(glmnet)
library(doParallel)
registerDoParallel(cores = 4) 
library(nestedcv)
## Set paths
proj_path <- "/ix/cigcore/proj/biswas"
data_path <- file.path(proj_path, "data", "Ascherman_TotalRNA_07_20_2023")
resPath <- gsub(pattern="data", replacement="result", data_path)
experiment <- "4-11-Lasso_WGCNA_16_logistic"
experimentPath <- file.path(resPath, experiment)
dir.create(experimentPath, showWarnings = TRUE, recursive = TRUE)
#metadata_path <- file.path(data_path, "metadata_remove_unpaired_low_align.csv")
#dir.create(webgestalt_path, recursive=TRUE)
metadata_path <- file.path(data_path, "all_metadata_remove_unpaired_low_align.csv") 
file_path <- file.path(data_path, "TPM_ordered_variance.csv")
```


```{r count, include=FALSE, echo=FALSE}
coldata <- read.csv(file.path(metadata_path), row.names = 1) #Metadata
coldata$sample <- rownames(coldata)  # new variable:sample = rownames in 'coldata'
#coldata <- coldata[ , c("sample", "treatment")]
coldata$treatment <- ifelse(coldata$treatment == "Pre", 0, 1)
coldata$treatment <- factor(coldata$treatment)
coldata$age_group <- factor(coldata$age_group)
TPM_ordered_variance <- read.csv(file.path(file_path), row.names = 1)
```

```{r, include=FALSE, echo=FALSE}
RSEM_path <- file.path(resPath, "rsem_counts")
rsem_files <- list.files(RSEM_path, pattern = "quant_star-rsem.genes.results", full.names = TRUE)
sample_names <- sapply(strsplit(basename(rsem_files), "_"), function(x) paste(x[1:4], collapse = "_"))
names(rsem_files) <- sample_names

txi.rsem <- tximport(rsem_files, type = "rsem", txIn = FALSE, txOut = FALSE) #txIn/Out = FALSE --> gene-level
# Still have 60708 genes (stay the same number after removing version number + duplicates)

#filter length: remove 0-length values
txi.rsem$abundance <-
  txi.rsem$abundance[apply(txi.rsem$length,
                                1,
                                function(row) all(row !=0 )),]
# keep counts have same number of levels as abundance 
txi.rsem$counts <-
  txi.rsem$counts[apply(txi.rsem$length,
                                1,
                                function(row) all(row !=0 )),]
# align samples: make sure rownames of txi.rsem correspond to sample names in coldata
counts <- as.data.frame(txi.rsem$counts)
counts <- counts[, rownames(coldata)]
# filter genes with sum of counts >= 10 in all samples 
# filtered_counts <- counts[rowSums(counts) >= 10, ] #29221

# remove genes with counts < 10 in over half of samples
remove_low_counts <- counts[rowSums(counts >= 10) >= 12, ] # 16941

nzv_genes <- nearZeroVar(t(remove_low_counts)) #didn't find any non-zero variance genes

# remove genes with a high frequency of zeros counts (0 count in over 20% of the samples)
remove_high_0_counts <- remove_low_counts[rowSums(remove_low_counts == 0) <= 0.2 * ncol(remove_low_counts), ] #16895

filtered_counts <- remove_high_0_counts

subset_abundance <- txi.rsem$abundance[rownames(filtered_counts), ]
tximport_TPM <- as.data.frame(subset_abundance) 
tximport_TPM <- tximport_TPM[, rownames(coldata)]

#check if there is any overlapping
#common_genes <- intersect(rownames(tximport_TPM), rownames(res_LRT))
#nrow(tximport_TPM)  #16895
#nrow(res_LRT) #19395
#length(common_genes) #16895

tximport_TPM$ensembl_gene_id_version <- rownames(tximport_TPM)
rownames(tximport_TPM) <- NULL
tximport_TPM$ensembl_gene_id <- sub("\\.\\d+$", "", tximport_TPM$ensembl_gene_id_version)
rownames(tximport_TPM) <- tximport_TPM$ensembl_gene_id

tximport_TPM <- tximport_TPM[, c(-25, -26)]
tximport_TPM <- tximport_TPM[(!duplicated(rownames(tximport_TPM))), ]

tximport_TPM[, 1:ncol(tximport_TPM)] <- sapply(tximport_TPM[, 1:ncol(tximport_TPM)], as.numeric) #convert it to numeric factors
all(colnames(tximport_TPM) %in% rownames(coldata)) # checking to see if sample names match between the two files
all(colnames(tximport_TPM) == rownames(coldata)) # checking to see if the sample names are in the same order in both files
TPM <- tximport_TPM
```

```{r, results = FALSE}
# TPM_ordered_variance_plot <- t(TPM_ordered_variance)
# qqnorm(t(tximport_TPM))
# abline(0, 1, col = "red") #TPM --> Poisson distribution
# 
# qqnorm(TPM_ordered_variance_plot) # --> normal distribution
# abline(0, 1, col = "red")
```

# Sample Metadata

```{r metadata, echo=FALSE}
sampleInfo <- coldata
numSamples <- dim(sampleInfo)[1]
sampleInfo$Sequence <- seq_len(nrow(sampleInfo))
rownames(sampleInfo) <- NULL
knitr::kable(sampleInfo)
```

# WGCNA result

> Soft threshold = 16

soft threshold = **16**
	
![](/ix/cigcore/proj/biswas/result/Ascherman_TotalRNA_07_20_2023/4-1-2023-WGCNA_SoftThreshold_16/thumbnail_image.png){width=60%}.

> Modules = 29

**29** modules in total

![](/ix/cigcore/proj/biswas/result/Ascherman_TotalRNA_07_20_2023/3-25-2023-WGCNA/module_16.png){width=70%}.



|  Positive modules | Spearman correlation (p-value) | 
| ------------- |:-------------:|
|  **lightgreen (152 genes)** | 0.14 (0.1)  | 


|  Negative modules | Spearman correlation (p-value) | 
| ------------- |:-------------:|
|  **darkred (63 genes)** | -0.12 (0.2)  | 
|  **midnightblue (303 genes)** | -0.1 (0.2) | 
 



```{r, out.width = '70%'}
knitr::include_graphics("/ix/cigcore/proj/biswas/result/Ascherman_TotalRNA_07_20_2023/4-1-2023-WGCNA_SoftThreshold_16/heatmap_SoftThreshold_16.png")
```


```{r, echo=FALSE}
#X: expression matrix
expression_matrix <- as.matrix(TPM_ordered_variance)
#Y: response variable 
response_variable <- coldata$treatment
response_variable <- as.factor(response_variable)
```

```{r, echo=FALSE}
file_path_lightgreen <- "/ix/cigcore/proj/biswas/result/Ascherman_TotalRNA_07_20_2023/3-25-2023-WGCNA/module_csv_SoftThreshold_16/lightgreen.csv"
file_path_darkred <- "/ix/cigcore/proj/biswas/result/Ascherman_TotalRNA_07_20_2023/3-25-2023-WGCNA/module_csv_SoftThreshold_16/darkred.csv"
file_path_midnightblue <- "/ix/cigcore/proj/biswas/result/Ascherman_TotalRNA_07_20_2023/3-25-2023-WGCNA/module_csv_SoftThreshold_16/midnightblue.csv"

positive <- read.csv(file_path_lightgreen) #152
negative_1 <- read.csv(file_path_darkred) #63
negative_2 <- read.csv(file_path_midnightblue) #303

positive_expression_matrix <- expression_matrix[,positive$gene] #lightgreen: 152
negative_expression_matrix_1 <- expression_matrix[,negative_1$gene] #darkred: 63
negative_expression_matrix_2 <- expression_matrix[,negative_2$gene] #midnightblue: 303
```

# Run LASSO on treatment-positive modules (Module lightgreen: 152 genes)

* Alpha = 1

* Nested cross validation

     + outer loop method: leave-one-out
     + inner loop method: leave-one-out
 
```{r}
  alpha <- 1
  
  set.seed(123)
  
 lassoModel <- glmnet(x = positive_expression_matrix, y = response_variable, alpha = alpha, family = "binomial", standardize = FALSE)

# plot(lassoModel, xvar="lambda")

 #summary(lassoModel)

  
  
   n_outer_folds <- length(response_variable)  # LOOCV
    n_inner_folds <- 23 # LOOCV
    outer_folds <- lapply(1:length(response_variable), function(i) i)
    
    set.seed(123)
    nested_cv_result <- nestcv.glmnet(
      y = response_variable, 
      x = positive_expression_matrix, 
      family = "binomial",
      outer_method = "LOOCV", 
      n_inner_folds = n_inner_folds,
      outer_folds = outer_folds, 
      alphaSet = 1,
      finalCV = TRUE
    )
    #cv.glmnet: pass "standardize TRUE" 
    
    
    # Get the lambda value, model and non-zero coefficients selected during cross-validation
    best_lambda <- nested_cv_result$final_param["lambda"]
    best_lambda_value <- as.numeric(best_lambda)
    current_model <- nested_cv_result$final_fit
    non_zero_coefficients <- rownames(nested_cv_result[["final_coef"]])[-1]
  
  cat("Tuned lambda value:\n", best_lambda_value, "\n")
  current_model
  cat("Non-zero Coefficients:\n", non_zero_coefficients, "\n")

  gene_positive <- non_zero_coefficients
  

```

## List of genes with non-zero coefficients (10 Genes)

```{r}

ensemblID <- gene_positive
attributes <- c("ensembl_gene_id", "external_gene_name")
ensemblMart <- useMart("ensembl", dataset="hsapiens_gene_ensembl")
external_gene_names_positive <- getBM(attributes = attributes, filters = "ensembl_gene_id", values = gene_positive, mart = ensemblMart)

knitr::kable(external_gene_names_positive, "html") %>%
  kableExtra::kable_styling(full_width = FALSE)

```

## AUC
```{r}
library(pROC)
proc <- nested_cv_result$roc
auc <- proc$auc
plot.roc(
  proc,
  grid = TRUE,
  grid.lwd = 2,
  col = "royalblue",
  legacy.axes = TRUE,
  main = paste("Nested-CV Lasso model (10 genes) in positive module (152 genes)\n Area under the curve (AUC):", round(auc, digits = 2))
)

```

## Accuracy
```{r}
nested_cv_result[["summary"]][["table"]]
nested_cv_result[["summary"]][["metrics"]]
```

## TPM (0: pre vs. 1: post)
```{r}
selected_genes <- external_gene_names_positive$ensembl_gene_id
gene_TPM <- TPM[rownames(TPM) %in% selected_genes, , drop = FALSE]
rownames(gene_TPM) <- external_gene_names_positive$external_gene_name
newcoldata <- coldata[,c(1,3)]
gene_TPM_treat <- merge(t(gene_TPM), newcoldata, by = "row.names")
names(gene_TPM_treat)[1] <- "sample"

knitr::kable(gene_TPM_treat, "html") %>%
  kableExtra::kable_styling(full_width = FALSE)
```

## Heatmap (Pre-treatment vs Post-treatment)


```{r}
library(RColorBrewer)
selected_genes <- external_gene_names_positive$ensembl_gene_id
zMatrix <- t(TPM_ordered_variance)
gene_zMatrix <- zMatrix[rownames(zMatrix) %in% selected_genes, , drop = FALSE]
gene_zMatrix <- as.data.frame(gene_zMatrix)

breaksList = seq(-3, 3, by = 0.06)

sorted <- coldata[order(coldata$treatment == '1', coldata$treatment), ]
cluster_treatment <- as.data.frame(t(sorted[, c(1:4)]))

gene_zMatrix_reordered <- gene_zMatrix[, match(names(cluster_treatment), names(gene_zMatrix))]


condition_reordered <- data.frame(sorted[, 1])
colnames(condition_reordered)[1] <- "treatment"
rownames(condition_reordered) <- colnames(gene_zMatrix_reordered)

gene_zMatrix_reordered <- gene_zMatrix_reordered[match(external_gene_names_positive$ensembl_gene_id, rownames(gene_zMatrix_reordered)), ]
rownames(gene_zMatrix_reordered) <- external_gene_names_positive$external_gene_name

p1 <- pheatmap(gene_zMatrix_reordered,
               color = colorRampPalette(rev(brewer.pal(n = 7, name =
  "RdYlBu")))(100),
               show_rownames = TRUE, 
               cluster_cols = FALSE, 
               cluster_rows = FALSE, 
               annotation_col = condition_reordered, 
               annotation_colors = list(treatment = c("0" = "lightpink", "1" = "lightgreen")),
               cutree_rows = 2, 
               cutree_cols = 2, 
               fontsize_row = 7,
               breaks = breaksList,
               legend = TRUE
)

p1

```

## Heatmap for Log-FoldChange for each patient

```{r}
selected_genes <- external_gene_names_positive$ensembl_gene_id
zMatrix <- t(TPM_ordered_variance)
gene_zMatrix <- zMatrix[rownames(zMatrix) %in% selected_genes, , drop = FALSE]
gene_zMatrix <- as.data.frame(t(gene_zMatrix))
df <- merge(as.data.frame(gene_zMatrix), as.data.frame(coldata), by = "row.names", all.x = TRUE)
df <- df[, c(2,3,4,5,6,7,8,9,10,11,12,14)]

df2 <- df %>%
  pivot_longer(cols = -c(patient_id, treatment), names_to = "Gene", values_to = "logTPM") %>%
  group_by(patient_id, Gene) 

treatment_1 <- df2 %>% dplyr::filter(treatment == 1)
treatment_0 <- df2 %>% dplyr::filter(treatment == 0)
merged_data <- merge(treatment_1, treatment_0, by = c("patient_id", "Gene"), suffixes = c("_treatment_1", "_treatment_0"))
merged_data <- merged_data %>%
  mutate(diff_logTPM = logTPM_treatment_1 - logTPM_treatment_0) %>%
  dplyr::select(Gene, patient_id, diff_logTPM) %>% 
  pivot_wider(names_from = Gene, values_from = diff_logTPM) 

rownames(merged_data) <- merged_data$patient_id
patient <- as.data.frame(unique(coldata$patient_id))
rownames(patient) <- patient$`unique(coldata$patient_id)`
merged_data_patinet_order <- merged_data[rownames(patient), ]

rownames(merged_data_patinet_order) <- merged_data_patinet_order$patient_id
gene_LFC <- as.data.frame(t(merged_data_patinet_order))
gene_LFC <- gene_LFC[-1, ]

gene_LFC_reordered <- gene_LFC[match(external_gene_names_positive$ensembl_gene_id, rownames(gene_LFC)), ]
rownames(gene_LFC_reordered) <- external_gene_names_positive$external_gene_name
#keep patient_id the same as the previous heatmap
breaksList = seq(-3, 3, by = 0.06)
p4 <- pheatmap(gene_LFC_reordered, 
         show_rownames = TRUE, 
         cluster_cols = FALSE, 
         cluster_rows = FALSE, 
         cutree_rows = 2, 
         cutree_cols = 2, 
         breaks = breaksList,
         fontsize_row = 7)
```

# Run LASSO on treatment-negative modules (Module darkred: 63 genes)

* Alpha = 1

* Nested cross validation

     + outer loop method: leave-one-out
     + inner loop method: k-fold; k = **4**
     + Didn't find any non-zero coeffients when using leave-one-out for inner loop
 
```{r}
coldata$reverse_treatment <- ifelse(coldata$treatment == 0, 1, 0)

# Converting coldata$reverse_treatment into a factor
reverse_response_variable <- as.factor(coldata$reverse_treatment)
  alpha <- 1
  
  set.seed(123)
  
  
   n_outer_folds <- length(reverse_response_variable)  # LOOCV
    n_inner_folds <- 4
    outer_folds <- lapply(1:length(reverse_response_variable), function(i) i)
    
    set.seed(123)
    nested_cv_result <- nestcv.glmnet(
      y = reverse_response_variable, 
      x = negative_expression_matrix_1, 
      family = "binomial",
      outer_method = "LOOCV", 
      n_inner_folds = n_inner_folds,
      outer_folds = outer_folds, 
      alphaSet = 1,
      finalCV = TRUE
    )
    
    # Get the lambda value, model and non-zero coefficients selected during cross-validation
    best_lambda <- nested_cv_result$final_param["lambda"]
    best_lambda_value <- as.numeric(best_lambda)
    current_model <- nested_cv_result$final_fit
    non_zero_coefficients <- rownames(nested_cv_result[["final_coef"]])[-1]
  
  cat("Tuned lambda value:\n", best_lambda_value, "\n")
  current_model
  cat("Non-zero Coefficients:\n", non_zero_coefficients, "\n")

  gene_negative_1 <- non_zero_coefficients
```

## AUC
```{r}
testy <- nested_cv_result[["output"]][["testy"]]
predyp <- nested_cv_result[["output"]][["predyp"]]
predyp <- 1-predyp
fit.roc <- pROC::roc(testy, predyp, direction = "<",
                         quiet = TRUE)

proc <- fit.roc
auc <- nested_cv_result$roc$auc
auc <- 1-auc
plot.roc(
  proc,
  grid = TRUE,
  grid.lwd = 2,
  col = "royalblue",
  legacy.axes = TRUE,
  main = paste("Nested-CV Lasso model (4 genes) in negative module (63 genes)\n Area under the curve (AUC):", round(auc, digits = 2))
)

```

## Accuracy
```{r}
library(caret)
cm <- confusionMatrix(table(fit.roc[["original.response"]], fit.roc[["response"]]))

cm[["table"]]
cm[["overall"]]
```

## List of genes with non-zero coefficients  (4 genes) 

```{r}
ensemblID <- gene_negative_1
attributes <- c("ensembl_gene_id", "external_gene_name")
#ensemblMart <- useMart("ensembl", dataset="hsapiens_gene_ensembl")
external_gene_names_negative_1 <- getBM(attributes = attributes, filters = "ensembl_gene_id", values = gene_negative_1, mart = ensemblMart)


external_gene_names_negative_1$external_gene_name <- ifelse(external_gene_names_negative_1$external_gene_name == "", 
                                                     external_gene_names_negative_1$ensembl_gene_id, 
                                                   external_gene_names_negative_1$external_gene_name)


knitr::kable(external_gene_names_negative_1, "html") %>%
  kableExtra::kable_styling(full_width = FALSE)
```

## TPM (0: pre vs. 1: post)

```{r}
selected_genes <- external_gene_names_negative_1$ensembl_gene_id
gene_TPM <- TPM[rownames(TPM) %in% selected_genes, , drop = FALSE]
rownames(gene_TPM) <- external_gene_names_negative_1$external_gene_name
newcoldata <- coldata[,c(1,3)]
gene_TPM_treat <- merge(t(gene_TPM), newcoldata, by = "row.names")
names(gene_TPM_treat)[1] <- "sample"

knitr::kable(gene_TPM_treat, "html") %>%
  kableExtra::kable_styling(full_width = FALSE)
```

## Heatmap (Pre-treatment vs Post-treatment)


```{r}
library(RColorBrewer)
selected_genes <- external_gene_names_negative_1$ensembl_gene_id
zMatrix <- t(TPM_ordered_variance)
gene_zMatrix <- zMatrix[rownames(zMatrix) %in% selected_genes, , drop = FALSE]
gene_zMatrix <- as.data.frame(gene_zMatrix)

breaksList = seq(-3, 3, by = 0.06)

sorted <- coldata[order(coldata$treatment == '1', coldata$treatment), ]
cluster_treatment <- as.data.frame(t(sorted[, c(1:4)]))

gene_zMatrix_reordered <- gene_zMatrix[, match(names(cluster_treatment), names(gene_zMatrix))]


condition_reordered <- data.frame(sorted[, 1])
colnames(condition_reordered)[1] <- "treatment"
rownames(condition_reordered) <- colnames(gene_zMatrix_reordered)

gene_zMatrix_reordered <- gene_zMatrix_reordered[match(external_gene_names_negative_1$ensembl_gene_id, rownames(gene_zMatrix_reordered)), ]
rownames(gene_zMatrix_reordered) <- external_gene_names_negative_1$external_gene_name

p1 <- pheatmap(gene_zMatrix_reordered,
               color = colorRampPalette(rev(brewer.pal(n = 7, name =
  "RdYlBu")))(100),
               show_rownames = TRUE, 
               cluster_cols = FALSE, 
               cluster_rows = FALSE, 
               annotation_col = condition_reordered, 
               annotation_colors = list(treatment = c("0" = "lightpink", "1" = "lightgreen")),
               cutree_rows = 2, 
               cutree_cols = 2, 
               fontsize_row = 7,
               breaks = breaksList,
               legend = TRUE
)

p1

```

## Heatmap for Log-FoldChange for each patient

```{r}
selected_genes <- external_gene_names_negative_1$ensembl_gene_id
zMatrix <- t(TPM_ordered_variance)
gene_zMatrix <- zMatrix[rownames(zMatrix) %in% selected_genes, , drop = FALSE]
gene_zMatrix <- as.data.frame(t(gene_zMatrix))
df <- merge(as.data.frame(gene_zMatrix), as.data.frame(coldata), by = "row.names", all.x = TRUE)
df <- df[, c(2,3,4,5,6,8)]

df2 <- df %>%
  pivot_longer(cols = -c(patient_id, treatment), names_to = "Gene", values_to = "logTPM") %>%
  group_by(patient_id, Gene) 

treatment_1 <- df2 %>% dplyr::filter(treatment == 1)
treatment_0 <- df2 %>% dplyr::filter(treatment == 0)
merged_data <- merge(treatment_1, treatment_0, by = c("patient_id", "Gene"), suffixes = c("_treatment_1", "_treatment_0"))
merged_data <- merged_data %>%
  mutate(diff_logTPM = logTPM_treatment_1 - logTPM_treatment_0) %>%
  dplyr::select(Gene, patient_id, diff_logTPM) %>% 
  pivot_wider(names_from = Gene, values_from = diff_logTPM) 

rownames(merged_data) <- merged_data$patient_id
patient <- as.data.frame(unique(coldata$patient_id))
rownames(patient) <- patient$`unique(coldata$patient_id)`
merged_data_patinet_order <- merged_data[rownames(patient), ]

rownames(merged_data_patinet_order) <- merged_data_patinet_order$patient_id
gene_LFC <- as.data.frame(t(merged_data_patinet_order))
gene_LFC <- gene_LFC[-1, ]

gene_LFC_reordered <- gene_LFC[match(external_gene_names_negative_1$ensembl_gene_id, rownames(gene_LFC)), ]
rownames(gene_LFC_reordered) <- external_gene_names_negative_1$external_gene_name
#keep patient_id the same as the previous heatmap
breaksList = seq(-3, 3, by = 0.06)
p4 <- pheatmap(gene_LFC_reordered, 
         show_rownames = TRUE, 
         cluster_cols = FALSE, 
         cluster_rows = FALSE, 
         cutree_rows = 2, 
         cutree_cols = 2, 
         breaks = breaksList,
         fontsize_row = 7)
```


# Run LASSO on treatment-negative module (Module midnightblue: 303 genes)

* Alpha = 1

* Nested cross validation


     + outer loop method: leave-one-out
     + inner loop method: k-fold; k = **4**
     + Didn't find any non-zero coeffients when using leave-one-out for inner loop
 
```{r}
  alpha <- 1
  
  set.seed(123)
  
  
   n_outer_folds <- length(reverse_response_variable)  # LOOCV
    n_inner_folds <- 4
    outer_folds <- lapply(1:length(reverse_response_variable), function(i) i)
    
    set.seed(123)
    nested_cv_result <- nestcv.glmnet(
      y = reverse_response_variable, 
      x = negative_expression_matrix_2, 
      family = "binomial",
      outer_method = "LOOCV", 
      n_inner_folds = n_inner_folds,
      outer_folds = outer_folds, 
      alphaSet = 1,
      finalCV = TRUE
    )
    
    # Get the lambda value, model and non-zero coefficients selected during cross-validation
    best_lambda <- nested_cv_result$final_param["lambda"]
    best_lambda_value <- as.numeric(best_lambda)
    current_model <- nested_cv_result$final_fit
    non_zero_coefficients <- rownames(nested_cv_result[["final_coef"]])[-1]
  
  cat("Tuned lambda value:\n", best_lambda_value, "\n")
  current_model
  cat("Non-zero Coefficients:\n", non_zero_coefficients, "\n")

  gene_negative_2 <- non_zero_coefficients
  
```

## AUC
```{r}
testy <- nested_cv_result[["output"]][["testy"]]
predyp <- nested_cv_result[["output"]][["predyp"]]
predyp <- 1-predyp
fit.roc <- pROC::roc(testy, predyp, direction = "<",
                         quiet = TRUE)

proc <- fit.roc
auc <- nested_cv_result$roc$auc
auc <- 1-auc
plot.roc(
  proc,
  grid = TRUE,
  grid.lwd = 2,
  col = "royalblue",
  legacy.axes = TRUE,
  main = paste("Nested-CV Lasso model (4 genes) in negative module (63 genes)\n Area under the curve (AUC):", round(auc, digits = 2))
)
```

## Accuracy
```{r}
cm <- confusionMatrix(table(fit.roc[["original.response"]], fit.roc[["response"]]))

cm[["table"]]
cm[["overall"]]
```

## List of genes with non-zero coefficients  (1 gene)

```{r}
ensemblID <- gene_negative_2
attributes <- c("ensembl_gene_id", "external_gene_name")
#ensemblMart <- useMart("ensembl", dataset="hsapiens_gene_ensembl")
external_gene_names_negative_2 <- getBM(attributes = attributes, filters = "ensembl_gene_id", values = gene_negative_2, mart = ensemblMart)


external_gene_names_negative_2$external_gene_name <- ifelse(external_gene_names_negative_2$external_gene_name == "", 
                                                     external_gene_names_negative_2$ensembl_gene_id, 
                                                   external_gene_names_negative_2$external_gene_name)


knitr::kable(external_gene_names_negative_2, "html") %>%
  kableExtra::kable_styling(full_width = FALSE)
```

## TPM (0: pre vs. 1: post) 

```{r}
selected_genes <- external_gene_names_negative_2$ensembl_gene_id
gene_TPM <- TPM[rownames(TPM) %in% selected_genes, , drop = FALSE]
rownames(gene_TPM) <- external_gene_names_negative_2$external_gene_name
newcoldata <- coldata[,c(1,3)]
gene_TPM_treat <- merge(t(gene_TPM), newcoldata, by = "row.names")
names(gene_TPM_treat)[1] <- "sample"

knitr::kable(gene_TPM_treat, "html") %>%
  kableExtra::kable_styling(full_width = FALSE)
```

## Heatmap (Pre-treatment vs Post-treatment)


```{r}
library(RColorBrewer)
selected_genes <- external_gene_names_negative_2$ensembl_gene_id
zMatrix <- t(TPM_ordered_variance)
gene_zMatrix <- zMatrix[rownames(zMatrix) %in% selected_genes, , drop = FALSE]
gene_zMatrix <- as.data.frame(gene_zMatrix)

breaksList = seq(-3, 3, by = 0.06)

sorted <- coldata[order(coldata$treatment == '1', coldata$treatment), ]
cluster_treatment <- as.data.frame(t(sorted[, c(1:4)]))

gene_zMatrix_reordered <- gene_zMatrix[, match(names(cluster_treatment), names(gene_zMatrix))]


condition_reordered <- data.frame(sorted[, 1])
colnames(condition_reordered)[1] <- "treatment"
rownames(condition_reordered) <- colnames(gene_zMatrix_reordered)

gene_zMatrix_reordered <- gene_zMatrix_reordered[match(external_gene_names_negative_2$ensembl_gene_id, rownames(gene_zMatrix_reordered)), ]
rownames(gene_zMatrix_reordered) <- external_gene_names_negative_2$external_gene_name

p1 <- pheatmap(gene_zMatrix_reordered,
               color = colorRampPalette(rev(brewer.pal(n = 7, name =
  "RdYlBu")))(100),
               show_rownames = TRUE, 
               cluster_cols = FALSE, 
               cluster_rows = FALSE, 
               annotation_col = condition_reordered, 
               annotation_colors = list(treatment = c("0" = "lightpink", "1" = "lightgreen")),
               cutree_rows = 2, 
               cutree_cols = 2, 
               fontsize_row = 7,
               breaks = breaksList,
               legend = TRUE
)

p1

```

## Heatmap for Log-FoldChange for each patient

```{r}
selected_genes <- external_gene_names_negative_2$ensembl_gene_id
zMatrix <- t(TPM_ordered_variance)
gene_zMatrix <- zMatrix[rownames(zMatrix) %in% selected_genes, , drop = FALSE]
gene_zMatrix <- as.data.frame(t(gene_zMatrix))
df <- merge(as.data.frame(gene_zMatrix), as.data.frame(coldata), by = "row.names", all.x = TRUE)
df <- df[, c(2,3,5)]

df2 <- df %>%
  pivot_longer(cols = -c(patient_id, treatment), names_to = "Gene", values_to = "logTPM") %>%
  group_by(patient_id, Gene) 

treatment_1 <- df2 %>% dplyr::filter(treatment == 1)
treatment_0 <- df2 %>% dplyr::filter(treatment == 0)
merged_data <- merge(treatment_1, treatment_0, by = c("patient_id", "Gene"), suffixes = c("_treatment_1", "_treatment_0"))
merged_data <- merged_data %>%
  mutate(diff_logTPM = logTPM_treatment_1 - logTPM_treatment_0) %>%
  dplyr::select(Gene, patient_id, diff_logTPM) %>% 
  pivot_wider(names_from = Gene, values_from = diff_logTPM) 

rownames(merged_data) <- merged_data$patient_id
patient <- as.data.frame(unique(coldata$patient_id))
rownames(patient) <- patient$`unique(coldata$patient_id)`
merged_data_patinet_order <- merged_data[rownames(patient), ]

rownames(merged_data_patinet_order) <- merged_data_patinet_order$patient_id
gene_LFC <- as.data.frame(t(merged_data_patinet_order))
gene_LFC <- gene_LFC[-1, ]

gene_LFC_reordered <- gene_LFC[match(external_gene_names_negative_2$ensembl_gene_id, rownames(gene_LFC)), ]
rownames(gene_LFC_reordered) <- external_gene_names_negative_2$external_gene_name
#keep patient_id the same as the previous heatmap
breaksList = seq(-3, 3, by = 0.06)
p4 <- pheatmap(gene_LFC_reordered, 
         show_rownames = TRUE, 
         cluster_cols = FALSE, 
         cluster_rows = FALSE, 
         cutree_rows = 2, 
         cutree_cols = 2, 
         breaks = breaksList,
         fontsize_row = 7)
```



# Final model
## Final Lasso nested cv (Total 15 genes)

```{r}
total_gene <- rbind(external_gene_names_positive, external_gene_names_negative_1, external_gene_names_negative_2)
```

```{r}
alpha <- 1

final_model_matrix <- as.data.frame(expression_matrix[, total_gene$ensembl_gene_id])
final_model_matrix <- merge(as.data.frame(final_model_matrix), as.data.frame(coldata[1]), by = "row.names", all.x = TRUE)
cv_matrix <- expression_matrix[, total_gene$ensembl_gene_id]

genes_final_model <- paste(total_gene$ensembl_gene_id, collapse = " + ")

formula_str <- paste("treatment ~", genes_final_model)

   n_outer_folds <- length(final_model_matrix$treatment)  # LOOCV
    n_inner_folds <- 23
    outer_folds <- lapply(1:length(response_variable), function(i) i)
    
    set.seed(123)
    final_lasso_nested_cv <- nestcv.glmnet(
      y = final_model_matrix$treatment, 
      x = cv_matrix, #'x' must be numeric
      family = "binomial",
      outer_method = "LOOCV", 
      n_inner_folds = n_inner_folds,
      outer_folds = outer_folds, 
      alphaSet = alpha,
      finalCV = TRUE
    )
    
    
    best_lambda <- final_lasso_nested_cv$final_param["lambda"]
    best_lambda_value <- as.numeric(best_lambda)
    current_model <- final_lasso_nested_cv$final_fit
    non_zero_coefficients <- rownames(final_lasso_nested_cv[["final_coef"]])[-1]
  
  cat("Tuned lambda value:\n", best_lambda_value, "\n")
  current_model
  cat("Non-zero Coefficients:\n", non_zero_coefficients, "\n")
  gene_final <- non_zero_coefficients
```

### AUC

```{r}
proc <- final_lasso_nested_cv$roc
auc <- proc$auc
plot.roc(
  proc,
  grid = TRUE,
  grid.lwd = 2,
  col = "royalblue",
  main = paste("Nested-CV Final model\n Area under the curve (AUC):", round(auc, digits = 2))
)

```

### Accuracy of nested cv

```{r}
final_lasso_nested_cv[["summary"]][["table"]]
final_lasso_nested_cv[["summary"]][["metrics"]]
```


## Final model (10 genes)

```{r}
final_model_matrix <- as.data.frame(expression_matrix[, gene_final])
final_model_matrix <- merge(as.data.frame(final_model_matrix), as.data.frame(coldata[1]), by = "row.names", all.x = TRUE)

genes_final_model <- paste(gene_final, collapse = " + ")

formula_str <- paste("treatment ~", genes_final_model)


finalModel <- glm(formula_str,
                  data = data.frame(final_model_matrix), #15 gene
                  family = binomial) #poisson distribution
                
#try linear link
summary(finalModel)
```

### Check correlation between each feature

```{r}
heatmap(cor(expression_matrix[, gene_final]))
```


## Check complete separation (perfect prediction)

```{r}
library("detectseparation")
formula_str <- paste("treatment ~", genes_final_model)
finalModel_sep <- glm(formula_str,
                  data = as.data.frame(final_model_matrix),
                  family = binomial(link = "logit"),
                  method = "detect_separation")
finalModel_sep
# Check for perfect separation
# the actual maximum likelihood estimates

# coef(finalModel) + coef(finalModel_sep)
# coef(summary(finalModel))[, "Std. Error"] + abs(coef(finalModel_sep))
```

## Method 1: Bayes

```{r}
library(arm)
#reference paper: Convergence Failures in Logistic Regression 
finalModel_bayes <- bayesglm(formula_str,
                  data = as.data.frame(final_model_matrix),
                  family = binomial(link = "logit"))
display(finalModel_bayes)

finalModel_bayes_sep <- bayesglm(formula_str,
                  data = as.data.frame(final_model_matrix),
                  family = binomial(link = "logit"),
                  method = "detect_separation")
finalModel_bayes_sep
```

### AUC:bayes

```{r}
require(pROC)
proc <- roc(final_model_matrix$treatment, fitted(finalModel_bayes), smooth=FALSE)
auc <-proc$auc
plot.roc(
  proc,
  grid = TRUE,
  grid.lwd = 2,
  col = "royalblue",
  legacy.axes = TRUE,
  main = paste("Final Lasso model\n Area under the curve (AUC):", round(auc, digits = 2))
)
```

## Method 2: Firth's Bias-Reduced Logistic Regression
Firth's bias reduction method, equivalent to penalization of the log-likelihood

```{r}
library(logistf)
firth_model <- logistf(formula_str,
                  data = as.data.frame(final_model_matrix))
summary(firth_model)


#check seperation
firth_model_sep <- logistf(formula_str,
                  data = as.data.frame(final_model_matrix),
                  method = "detect_separation")
firth_model_sep
```

```{r}
predictions <- predict(firth_model, newdata = final_model_matrix, type = "response")

#AUC
auc_obj <- prediction(predictions, final_model_matrix$treatment)
auc_value <- as.numeric(performance(auc_obj, "auc")@y.values)
print(paste("AUC (test):", auc_value))

proc_test <- roc(final_model_matrix$treatment,predictions)
auc_test <-proc_test$auc
plot.roc(
  proc_test,
  grid = TRUE,
  grid.lwd = 2,
  col = "royalblue",
  main = paste("Final Lasso model on test data\n Area under the curve (AUC):", round(auc_test, digits = 2))
)

# accuracy
predicted_classes <- ifelse(predictions > 0.5, 1, 0)
accuracy <- sum(predicted_classes == final_model_matrix$treatment) / length(final_model_matrix$treatment)
print(paste("Accuracy (test):", accuracy))
```



Check direction of each gene in two models, all the same
```{r}
final_gene_matrix <- total_gene[total_gene$ensembl_gene_id %in% gene_final, ]
```

```{r}
coefficients_bayes <- as.data.frame(finalModel_bayes[["coefficients"]])
coefficients_bayes$ensembl_gene_id <- rownames(coefficients_bayes)
colnames(coefficients_bayes)[1] <- "bayes"

coefficients_firth <- as.data.frame(firth_model[["coefficients"]])
coefficients_firth$ensembl_gene_id <- rownames(coefficients_firth)
colnames(coefficients_firth)[1] <- "firth"

merge_coefficients <- merge(coefficients_bayes, coefficients_firth, by = "ensembl_gene_id")
merge_coefficients <- merge_coefficients[-1,]
merge_coefficients <- merge(final_gene_matrix, merge_coefficients, by = "ensembl_gene_id", all = TRUE)


merge_coefficients$direction_match <- ifelse(sign(merge_coefficients$bayes) == sign(merge_coefficients$firth), "Same", "Different")



```

# Final genes list (Total 10 genes)


| Modules (size)       | Module correlation to treatment     | Genes selected by lasso |
| ------------- |:-------------:| -----:|
|  **lightgreen (152 genes)**      | Positive | 8 |
|  **darkred (63 genes)**      | Negative      |   1 |
| **midnightblue (303 genes)**      | Negative      |    1 |

```{r}
total_gene <- rbind(external_gene_names_positive, external_gene_names_negative_1, external_gene_names_negative_2)
final_gene_matrix <- total_gene[total_gene$ensembl_gene_id %in% gene_final, ]
rownames(final_gene_matrix) <- seq_len(nrow(final_gene_matrix))

knitr::kable(final_gene_matrix, "html",row.names = TRUE) %>%
  kableExtra::kable_styling(full_width = FALSE)
```
  
## Heatmap (Pre-treatment vs Post-treatment)


```{r}
library(RColorBrewer)
selected_genes <- final_gene_matrix$ensembl_gene_id
zMatrix <- t(TPM_ordered_variance)
gene_zMatrix <- zMatrix[rownames(zMatrix) %in% selected_genes, , drop = FALSE]
gene_zMatrix <- as.data.frame(gene_zMatrix)

breaksList = seq(-3, 3, by = 0.06)

sorted <- coldata[order(coldata$treatment == '1', coldata$treatment), ]
cluster_treatment <- as.data.frame(t(sorted[, c(1:4)]))

gene_zMatrix_reordered <- gene_zMatrix[, match(names(cluster_treatment), names(gene_zMatrix))]


condition_reordered <- data.frame(sorted[, 1])
colnames(condition_reordered)[1] <- "treatment"
rownames(condition_reordered) <- colnames(gene_zMatrix_reordered)

gene_zMatrix_reordered <- gene_zMatrix_reordered[match(final_gene_matrix$ensembl_gene_id, rownames(gene_zMatrix_reordered)), ]
rownames(gene_zMatrix_reordered) <- final_gene_matrix$external_gene_name

p1 <- pheatmap(gene_zMatrix_reordered,
               color = colorRampPalette(rev(brewer.pal(n = 7, name =
  "RdYlBu")))(100),
               show_rownames = TRUE, 
               cluster_cols = FALSE, 
               cluster_rows = FALSE, 
               annotation_col = condition_reordered, 
               annotation_colors = list(treatment = c("0" = "lightpink", "1" = "lightgreen")),
               cutree_rows = 2, 
               cutree_cols = 2, 
               fontsize_row = 7,
               breaks = breaksList,
               legend = TRUE
)

p1

```

## Heatmap for Log-FoldChange for each patient

```{r}
selected_genes <- final_gene_matrix$ensembl_gene_id
zMatrix <- t(TPM_ordered_variance)
gene_zMatrix <- zMatrix[rownames(zMatrix) %in% selected_genes, , drop = FALSE]
gene_zMatrix <- as.data.frame(t(gene_zMatrix))
df <- merge(as.data.frame(gene_zMatrix), as.data.frame(coldata), by = "row.names", all.x = TRUE)
df <- df[, c(2,3,4,5,6,7,8,9,10,11,12,14)]

df2 <- df %>%
  pivot_longer(cols = -c(patient_id, treatment), names_to = "Gene", values_to = "logTPM") %>%
  group_by(patient_id, Gene) 

treatment_1 <- df2 %>% dplyr::filter(treatment == 1)
treatment_0 <- df2 %>% dplyr::filter(treatment == 0)
merged_data <- merge(treatment_1, treatment_0, by = c("patient_id", "Gene"), suffixes = c("_treatment_1", "_treatment_0"))
merged_data <- merged_data %>%
  mutate(diff_logTPM = logTPM_treatment_1 - logTPM_treatment_0) %>%
  dplyr::select(Gene, patient_id, diff_logTPM) %>% 
  pivot_wider(names_from = Gene, values_from = diff_logTPM) 

rownames(merged_data) <- merged_data$patient_id
patient <- as.data.frame(unique(coldata$patient_id))
rownames(patient) <- patient$`unique(coldata$patient_id)`
merged_data_patinet_order <- merged_data[rownames(patient), ]

rownames(merged_data_patinet_order) <- merged_data_patinet_order$patient_id
gene_LFC <- as.data.frame(t(merged_data_patinet_order))
gene_LFC <- gene_LFC[-1, ]

gene_LFC_reordered <- gene_LFC[match(final_gene_matrix$ensembl_gene_id, rownames(gene_LFC)), ]
rownames(gene_LFC_reordered) <- final_gene_matrix$external_gene_name
#keep patient_id the same as the previous heatmap
breaksList = seq(-3, 3, by = 0.06)
p4 <- pheatmap(gene_LFC_reordered, 
         show_rownames = TRUE, 
         cluster_cols = FALSE, 
         cluster_rows = FALSE, 
         cutree_rows = 2, 
         cutree_cols = 2, 
         breaks = breaksList,
         fontsize_row = 7)
```

# Pathway Analysis ORA

```{r ora1, echo=FALSE, warning=FALSE, message=FALSE, include=FALSE}
library(clusterProfiler) # perform ORA
library(org.Hs.eg.db) # Homo sapiens annotation --> gene identifier conversion
library(ReactomePA) # Reactome

#Each gene is represented by a human Entrez gene ID, which is the default keytype used by the clusterProfiler functions (and the only keytype compatible with ReactomePA::enrichPathway).
cp_path <- file.path(experimentPath, "pathway_analysis_ORA")
dir.create(cp_path, recursive=TRUE)

genes <- final_gene_matrix$ensembl_gene_id
entrez_genes <- bitr(genes, fromType = "ENSEMBL", toType = "ENTREZID", OrgDb = "org.Hs.eg.db")

# Reactome pathway over-representation analysis
    reactome_1 <- enrichPathway(gene = entrez_genes$ENTREZID, organism = "human", pvalueCutoff=0.2, pAdjustMethod = "BH", qvalueCutoff = 0.2, readable=TRUE)
    ORA <- list()
    ORA[[paste0("Reactome", "_WGCNA_16_logistic")]] <- reactome_1
    tab.reactome_1 <- as.data.frame(reactome_1)
    
# KEGG pathway over-representation analysis
    kegg_1 <- enrichKEGG(gene = entrez_genes$ENTREZID, organism = "hsa",  pvalueCutoff=0.2, pAdjustMethod = "BH", qvalueCutoff = 0.2, use_internal_data = FALSE)
    # mapping geneID to gene Symbol
    kegg_gene_1 <- setReadable(kegg_1, OrgDb = "org.Hs.eg.db", keyType="ENTREZID")
    kegg_gene_1@result$Description <- gsub(pattern = " - Homo sapiens (human)", replacement = "", kegg_gene_1@result$Description, fixed = T)    
    ORA[[paste0("KEGG", "_WGCNA_16_logistic")]] <- kegg_gene_1
    tab.kegg_1 <- as.data.frame(kegg_gene_1)

    plots_list <- list()
    
for(a1 in names(ORA)){
  reacORkegg <- ORA[[a1]]
  if(nrow(reacORkegg) !=0){
    # Create directories for each result
    ORA_path <- file.path(cp_path, a1)
    dir.create(ORA_path, recursive=TRUE)
    
    ORA_df <- as.data.frame(reacORkegg)
    write.csv(ORA_df, file.path(ORA_path, paste0(a1, ".csv")))
    
    # Plot enrichment result - dotplot
    fit <- plot(dotplot(reacORkegg, showCategory = 40))
    png(filename = file.path(ORA_path, paste0(a1, "_dotplot.png")), res = 300, width = 17, height = 10, units = "in")
    print(fit)
    dev.off()
    
    plots_list[[a1]] <- fit
  }
  
  if (nrow(reacORkegg) > 20){ #change number
    # Weighted Set Cover of geneSets
    weightedPath <- file.path(cp_path, "weightedSetCover", a1)
    dir.create(weightedPath, recursive=TRUE)
    
    setCoverNum = abs(0.40*(nrow(ORA_df)))
    nThreads = 4
    idsInSet <- sapply(ORA_df$geneID, strsplit, split="/")
    names(idsInSet) <- ORA_df$ID    
    minusLogP <- -log(ORA_df$pvalue)
    minusLogP[minusLogP == Inf] <- -log(.Machine$double.eps)
    wscRes <- weightedSetCover(idsInSet=idsInSet, costs=(1 / minusLogP), topN=setCoverNum, nThreads=nThreads)

    wscRes_full <- ORA_df[c(match(wscRes$topSets, ORA_df$ID)),]
    wscRes_full <- wscRes_full[order(wscRes_full$p.adjust), ]
    write.csv(wscRes_full, file.path(weightedPath, paste0(a1, ".csv")))
    
    # Plot enrichment result of weighted set cover - dotplot
    reacORkegg2 <- reacORkegg
    reacORkegg2@result <- wscRes_full
      
    # Set larger bottom margin to make space for y-axis labels
    fit <- plot(dotplot(reacORkegg2, showCategory = 40), main = paste0(a1))
    png(filename = file.path(weightedPath, paste0(a1, "_dotplot.png")), res = 300, width = 17, height = 28, units = "in")
    print(fit)
    dev.off()
    plots_list[[paste0(a1, "weightSet")]] <- fit
  }
}    
```

> Reactome

![](/ix/cigcore/proj/biswas/result/Ascherman_TotalRNA_07_20_2023/4-11-Lasso_WGCNA_16_logistic/pathway_analysis_ORA/weightedSetCover/Reactome_WGCNA_16_logistic/Reactome_WGCNA_16_logistic_dotplot.png).

> KEGG

![](/ix/cigcore/proj/biswas/result/Ascherman_TotalRNA_07_20_2023/4-11-Lasso_WGCNA_16_logistic/pathway_analysis_ORA/KEGG_WGCNA_16_logistic/KEGG_WGCNA_16_logistic_dotplot.png).

